src/utils/SquareGrid.h

```diff
@@ -0,0 +1,117 @@
+class SquareGrid
+{
+public:
+
+    /*! \brief Constructs a grid with the specified cell size.
+     *
+     * \param[in] cell_size The size to use for a cell (square) in the grid.
+     */
+    SquareGrid(coord_t cell_size);
+
+    coord_t getCellSize() const;
+
+    using GridPoint = Point;
+    using grid_coord_t = coord_t;
+
+    /*! \brief Process cells along a line indicated by \p line.
+     *
+     * \param[in] line The line along which to process cells
+     * \param[in] process_func Processes each cell.  process_func(elem) is
+     *    called for each cell. Processing stops if function returns false.
+     * \return Whether we need to continue processing after this function
+     */
+    bool processLineCells(const std::pair<Point, Point> line,
+                         const std::function<bool (GridPoint)>& process_cell_func);
+
+    /*! \brief Process cells along a line indicated by \p line.
+     *
+     * \param[in] line The line along which to process cells
+     * \param[in] process_func Processes each cell.  process_func(elem) is
+     *    called for each cell. Processing stops if function returns false.
+     * \return Whether we need to continue processing after this function
+     */
+    bool processLineCells(const std::pair<Point, Point> line,
+                         const std::function<bool (GridPoint)>& process_cell_func) const;
+
+    /*!
+     * process all cells in between the line from,to and (to.x,from.y),to
+     */
+    bool processAxisAlignedTriangle(const Point from, const Point to, const std::function<bool (GridPoint)>& process_cell_func) const;
+
+    bool processAxisAlignedTriangle(const Point from, const Point to, bool to_the_right, const std::function<bool (GridPoint)>& process_cell_func) const;
+
+    /*! \brief Process cells that might contain sought after points.
+    * 
+    * Processes cells that might be within \p radius of \p query_pt.
+    * May process elements that are up to radius + cell_size from query_pt.
+    *
+    * \param[in] query_pt The point to search around.
+    * \param[in] radius The search radius.
+    * \param[in] process_func Processes each cell.  process_func(loc) is
+    *    called for each cell coord. Processing stops if function returns false.
+    * \return Whether we need to continue processing after this function
+    */
+    bool processNearby(const Point &query_pt, coord_t radius,
+                       const std::function<bool (const GridPoint&)>& process_func) const;
+
+    /*! \brief Compute the grid coordinates of a point.
+     *
+     * \param[in] point The actual location.
+     * \return The grid coordinates that correspond to \p point.
+     */
+    GridPoint toGridPoint(const Point& point) const;
+
+    /*! \brief Compute the grid coordinate of a print space coordinate.
+     *
+     * \param[in] coord The actual location.
+     * \return The grid coordinate that corresponds to \p coord.
+     */
+    grid_coord_t toGridCoord(const coord_t& coord) const;
```

Comment:
Why not make this const?

Reply:
Parameter `coord` is const-qualified in the function declaration; const-qualification of parameters only has an effect in function definitions (clang-tidy)

Reply:
I think repeating that it's const is better. That way you can easily verify the const-correctness when you're scrolling through the code.

Reply:
I don't feel that strongly about this subject. But I do think static code analysis warnings are best solved. For the obvious reasons that it reduces risk but more importantly that you don't overlook other more important warnings because of the noise of these warnings.

Reply:
You should be able to disable (or enable!) specific warnings though, right?

Reply:
yes, if we have a .clang-tidy file in the root of the project. Which I introduced in the conan branch https://github.com/Ultimaker/CuraEngine/blob/conan/.clang-tidy